File: day04/solution2.rs
// Advent of Code 2021 Day 4 part 2
// Written by Jacob Morris (jacobm3@vt.edu)

use std::fs::File; // File I/O
use std::env; // Read command line args
use std::io::{prelude::*, BufReader};

static BOARD_SIZE: usize = 5;

/// A Bingo Board object. It stores the bingo board numbers as well as a mask of whether each
/// number has been drawn
#[derive(Debug)]
struct BingoBoard {
    board: Vec<Vec<u32>>,
    drawn: Vec<Vec<bool>>,
    bingo: bool,
}

/// Bingo board methods
impl BingoBoard {
    /// Check if this bingo board has a bingo
    fn check_bingo(&mut self) -> bool {
        for row in &self.drawn {
            if row.iter().all(|x| *x == true) {
                //println!("{}", board_index);
                self.bingo = true;
            }
        }
        //let tr = transpose(self.drawn);
        for col in transpose(self.drawn.clone()) {
            if col.iter().all(|x| *x == true) {
                //println!("{}", board_index);
                self.bingo = true;
            }
        }
        return self.bingo;
    }

    /// Sums all the unmarked numbers on the board and returns the product of that sum and the
    /// given last number
    fn score_board(&self, last_number: u32) -> u32 {
        let mut sum = 0;
        for row in 0..self.size() {
            for col in 0..self.size() {
                if self.drawn[row][col] == false {
                    sum += self.board[row][col];
                }
            }
        }
        return sum * last_number;
    }

    /// checks if the given number is on the board. If it is, then we set that mask value to true.
    fn check_board_for_number(&mut self, number: u32) -> () {
        for row in 0..self.size() {
            for col in 0..self.size() {
                if self.board[row][col] == number {
                    self.drawn[row][col] = true;
                }
            }
        }
    }

    /// Returns the size of the board. Note that a bingo board is a square so the height and width
    /// are equal
    fn size(&self) -> usize {
        return self.board.len();
    }
}


fn read_input(filename: String) -> (Vec<u32>, Vec<BingoBoard>) {
    println!("Reading input from {}", filename);
    // open file
    let file = File::open(filename).expect("file not found");
    let mut buf = BufReader::new(file);

    let mut first_line = String::new();
    // Read the first line containing the numbers that are pulled
    buf.read_line(&mut first_line);
    // remove \n
    first_line.pop();
    let draw_order = first_line
        .split(',')
        .map(|x| x.parse::<u32>().unwrap())
        .collect::<Vec<u32>>();

    // store all the found bingo boards
    let mut all_boards: Vec<BingoBoard> = Vec::new();
    let drawn = vec![vec![false; BOARD_SIZE]; BOARD_SIZE];

    // store the current board.
    let mut curr_board: Vec<Vec<u32>> = Vec::new();
    // parse the remaining boards
    for line in buf.lines().skip(1) {
        let line = line.expect("Could not read line.");
        // check if we are on an empty line and skip it.
        if line == "" {
            continue;
        }
        let row = line
            .split_whitespace()
            .map(|x| x.parse::<u32>().unwrap())
            .collect::<Vec<u32>>();
        // populate board
        curr_board.push(row);
        // board has been filled, empty it's contents
        if curr_board.len() == BOARD_SIZE {
            // make current bingo board
            let bb = BingoBoard {
                board: curr_board.clone(),
                drawn: drawn.clone(),
                bingo: false,
            };

            // Add current board to the list of all boards
            all_boards.push(bb);
            // clean up for new board
            curr_board.clear();
        }
    }
    return (draw_order, all_boards);
}

// Transpose vector of vectors:
// https://stackoverflow.com/questions/64498617/how-to-transpose-a-vector-of-vectors-in-rust
fn transpose<T>(v: Vec<Vec<T>>) -> Vec<Vec<T>>
where
    T: Clone,
{
    assert!(!v.is_empty());
    (0..v[0].len())
        .map(|i| v.iter().map(|inner| inner[i].clone()).collect::<Vec<T>>())
        .collect()
}

/// main driver for bingo code
fn play_bingo(draw_order: Vec<u32>, boards: &mut Vec<BingoBoard>) -> u32 {
    // keep track of which boards have won
    let mut losing_boards: Vec<usize> = (0..boards.len()).collect();
    for (ni, number) in draw_order.iter().enumerate() {
        println!("draw: {}", number);
        // Store all the boards that have won
        let mut winning_board_indices: Vec<usize> = Vec::new();
        // iterate over all boards and update their masks based on the number drawn
        let mut board_index = 0;
        for board in &mut *boards {
            // updates the board with the numbers drawn
            board.check_board_for_number(*number);
            // check if current board has won -- this updates it's bingo attribute
            if board.check_bingo() {
                // return the index of the winning board so we can remove it from the losing_boards
                // vector
                let winning_board_index = match losing_boards
                    .iter().position(|x| *x == board_index) {
                        Some(x) => Some(x),
                        None => None,
                };
                // remove the board from losing_boards vector
                if winning_board_index != None {
                    losing_boards.remove(winning_board_index.unwrap());
                }
            }
            // we have a losing board break out of this loop and finish that board
            if losing_boards.len() == 1 {
                break;
            }
            board_index += 1;
        }
        // finish the bingo game for the final losing board
        if losing_boards.len() == 1 {
            // get the board as a mutable reference.
            let mut losing_board = &mut boards[losing_boards[0]];
            // iterate over the rest of the number
            for num in draw_order.iter().skip(ni) {
                losing_board.check_board_for_number(*num);
                if losing_board.check_bingo() {
                    return losing_board.score_board(*num);
                }
            }
        }
    }
    return 0;
}

fn main() {
    // Read in command line arguments boilerplate
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Please provide an input file... \nNone provided - exiting");
        std::process::exit(-1);
    }
    let filename = &args[1];
    // end boilerplate

    // don't use & so that filename is cleaned up
    let (draw_order, mut boards) = read_input(String::from(filename));
    // play the bingo game
    let answer: u32 = play_bingo(draw_order, &mut boards);

    // print answer
    println!("The answer is {}", answer)
}

File: day04/solution2.py
"""
Implementation for AOC 2021 Day 4 Problem 2
"""

from typing import List, Tuple
import sys
import numpy as np
import ipdb

BOARD_WIDTH = 5
BOARD_HEIGHT = 5


def read_input(filename: str) -> Tuple[List[int], np.ndarray]:
    """
    Returns a 2D array of the bits
    """
    with open(filename) as f:
        lines = f.readlines()

    draw_order = [int(v) for v in lines[0].strip().split(',')]
    # number of board = len(lines[2:]) // 6
    n_boards = len(lines[1:]) // 6
    # init boards
    boards = np.zeros((n_boards, BOARD_WIDTH, BOARD_HEIGHT), dtype=int)
    for i in range(n_boards):
        # get rid of the newline row -- idk what's happening here
        board = lines[i*(BOARD_HEIGHT + 1) + 2: i*(BOARD_HEIGHT + 1) + (2 + BOARD_WIDTH)]
        new_board = np.zeros((BOARD_WIDTH, BOARD_HEIGHT), dtype=int)
        # iterate ove all rows and convert values to int
        for i_row, row in enumerate(board):
            new_row = [v for v in row.strip().split(' ') if v != '']
            # set the new row
            new_board[i_row, :] = new_row
        boards[i] = new_board
    return draw_order, boards


def check_for_winner(boards: np.ndarray, board_masks: np.ndarray) -> np.ndarray:
    """
    Returns whether each board has won or not
    """
    winning_boards = np.zeros((len(boards)), dtype=bool)
    for board_index, board in enumerate(board_masks):
        for row in board:
            if (row is True).all():
                winning_boards[board_index] = True
        for col in board.T:
            if (col is True).all():
                winning_boards[board_index] = True
    return winning_boards


def solve(draw_order, boards: np.ndarray) -> int:
    """
    """
    # Keep track of what numbers have been called
    board_masks = np.zeros_like(boards, dtype=bool)

    for draw_idx, drawn_number in enumerate(draw_order):
        for i_board, board in enumerate(boards):
            ixs = np.where(board == drawn_number)
            # ipdb.set_trace()
            if ixs is not None:
                board_masks[i_board][ixs[0], ixs[1]] = True

        # check if someone has won
        winning_boards = check_for_winner(boards, board_masks)
        # Get all the losing boards
        losing_board_indices = np.arange(len(winning_boards))[winning_boards is False]
        if len(losing_board_indices) == 1:
            # Get the final remaining board
            losing_board = boards[losing_board_indices]
            losing_board_mask = board_masks[losing_board_indices]
            # iterate through the rest of the numbers until this board wins
            # THIS IS DONE POORLY
            print(f"draw: {drawn_number}")
            for drawn_number in draw_order[draw_idx:]:
                ixs = np.where(losing_board[0] == drawn_number)
                print(f"draw: {drawn_number}")
                if ixs is not None:
                    losing_board_mask[0][ixs[0], ixs[1]] = True

                final_win = check_for_winner(losing_board, losing_board_mask)
                if final_win:
                    unmarked_numbers = losing_board[losing_board_mask is False]
                    return np.sum(unmarked_numbers) * drawn_number


def main():
    draw_order, boards = read_input(sys.argv[1])
    answer = solve(draw_order, boards)
    print(f"The answer is: {answer}")


if __name__ == "__main__":
    main()
